rfd/0194-selinux-policy.md

```diff
@@ -0,0 +1,150 @@
+
+4 main SELinux domains will be created to ensure each process in the chain is granted only the necessary permissions: `teleport_ssh_service`,
+`teleport_ssh_service_intermediary`, `teleport_ssh_service_session`, and `teleport_ssh_service_networking` respectively.
+
+Configurable file and port contexts will be created for the `teleport_ssh_service` domain, and configurable file contexts
+will be created for the `teleport_ssh_service_intermediary` domain to allow those processes to bind to the necessary ports and read and write
+to the necessary files, depending on the SSH agent configuration. At a minimum this will allow the `teleport_ssh_service` domain to bind to
+the configured SSH port and read and write to the data directory, and dial the configured Teleport Proxy and Auth servers. 
+
+There are a few optional features that can be enabled or disabled on SSH agents, the first version of the policy will ensure necessary permissions
+for all optional features are allowed to simplify development and deployment.
+
+These include:
+
+- Reading `~/.tsh/environment` on the server before creating a session
+    - The `teleport_ssh_service_intermediary` domain will be allowed to read `.tsh/environment` in all users' home directories
+- Creation of host users and groups
+    - The `teleport_ssh_service` domain will be allowed to call necessary binaries such as `useradd`, `usermod` and `groupadd`
+- Enhanced session recording with BPF
+   - The `teleport_ssh_service` domain will be allowed to create and manage BPF ring buffers, and load BPF programs
+- Creation of PAM contexts
+    - The `teleport_ssh_service_intermediary` domain will be allowed to `dlopen` the PAM module and call it's exported functions
+    - The location of the PAM module will be configurable
+- TCP port forwarding
+    - The `teleport_ssh_service_networking` domain will be allowed to listen to arbitrary ports and connect to arbitrary network addresses
+- X11 forwarding
+    - The `teleport_ssh_service_networking` domain will be allowed to create and manage X11 sockets
+- SFTP support
+    - The `teleport_ssh_service_intermediary` domain will be allowed to re-execute itself with `sftp` as the first argument
+
+The SELinux policy will be created as a SELinux Reference Policy (refpolicy), to maximize readability and maintainability.
+
+The main alternative is to use SELinux Common Intermediate Language (CIL) which bridges refpolicies and compiled binary policy files.
+CIL is extremely powerful but much more difficult to understand by humans, it's meant for consumption by SELinux tools.
+
+#### Future updates
+
+The SELinux policy will be updated in the future to support new features and maintain compatibility with existing features.
+Any changes to the SSH service that result in new files being managed, new network activity (listening on new ports, dialing new addresses etc),
+new binaries being executed, or new Linux syscalls being called will require updates to the SELinux policy.
```

Comment:
IMO Teleport should have a command that collects relevant audit log information, and I think that this should be a part of the initial release. This could be pretty simple - something logically along the lines of `cat /var/log/audit/audit.log | grep AVC | grep teleport > /var/run/teleport/support.log`. SELinux policy issues can be a pain to track down, especially without audit logs. This would help catch regressions, and add support in debugging problems.

Passing this through `audit2why` or `audit2allow` could help as well, _if_ they are available on client systems.

Reply:
+1 for the troublehsooting command as we'll receive support requests about this, or will have to rule out SELinux when debugging issues on SELinux-enforcing systems. Being able to easily extract traces saves a lot of time for customers, the support frontline, and core engineers.

Reply:
That's a good idea, and shouldn't be hard to add if we're just wrapping other commands.